### PR TITLE
[SPARK-9065][Streaming][PySpark] Add MessageHandler for Kafka Python API

### PR DESCRIPTION
**Pull request by zsxwing**
Fixed the merge conflicts in #7410

Closes #7410

### CODE REVIEW

**Comment by zsxwing** [comment]
@jerryshao After discussing with @tdas offline, I added some codes to speed up the case when the user doesn't provide a messageHandler. So that we can avoid message serialization for this case and won't reduce the performance. Could you take a look?

**Comment by jerryshao** [comment]
OK, let me check.

**Comment by jerryshao** [comment]
LGTM.

**Comment by zsxwing** [comment]
retest this please

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 1]
```diff
@@ -566,86 +579,97 @@ private[kafka] class KafkaUtilsPythonHelper {
...
       kafkaParams: JMap[String, String],
       offsetRanges: JList[OffsetRange],
-      leaders: JMap[TopicAndPartition, Broker]): JavaPairRDD[Array[Byte], Array[Byte]] = {
-    val messageHandler = new JFunction[MessageAndMetadata[Array[Byte], Array[Byte]],
-      (Array[Byte], Array[Byte])] {
-      def call(t1: MessageAndMetadata[Array[Byte], Array[Byte]]): (Array[Byte], Array[Byte]) =
-        (t1.key(), t1.message())
-    }
+      leaders: JMap[TopicAndPartition, Broker]
+      ): JavaRDD[(Array[Byte], Array[Byte])] = {
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: nit: incorrect formatting.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 2]
```diff
@@ -566,86 +579,97 @@ private[kafka] class KafkaUtilsPythonHelper {
...
+    val messageHandler =
+      (mmd: MessageAndMetadata[Array[Byte], Array[Byte]]) => (mmd.key, mmd.message)
+    new JavaRDD(createRDD(jsc, kafkaParams, offsetRanges, leaders, messageHandler))
+  }
+
+  def createRDDWithMessageHandler(
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: this should not be public.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala [hunk 3]
```diff
@@ -566,86 +579,97 @@ private[kafka] class KafkaUtilsPythonHelper {
...
   }
 
-  def createDirectStream(
+  def createDirectStreamWithMessageHandler(
       jssc: JavaStreamingContext,
       kafkaParams: JMap[String, String],
       topics: JSet[String],
       fromOffsets: JMap[TopicAndPartition, JLong]
-    ): JavaPairInputDStream[Array[Byte], Array[Byte]] = {
+      ): JavaDStream[Array[Byte]] = {
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala]: incorrect indent

File: python/pyspark/streaming/kafka.py [hunk 4]
```diff
@@ -19,12 +19,14 @@
...
 from pyspark.storagelevel import StorageLevel
-from pyspark.serializers import PairDeserializer, NoOpSerializer
+from pyspark.serializers import AutoBatchedSerializer, PickleSerializer, PairDeserializer, \
+    NoOpSerializer
 from pyspark.streaming import DStream
 from pyspark.streaming.dstream import TransformedDStream
 from pyspark.streaming.util import TransformFunction
 
-__all__ = ['Broker', 'KafkaUtils', 'OffsetRange', 'TopicAndPartition', 'utf8_decoder']
+__all__ = ['Broker', 'default_message_handler', 'KafkaMessageAndMetadata', 'KafkaUtils',
```
Comment (jerryshao) [inline python/pyspark/streaming/kafka.py]: Remove this `default_message_handler`

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
All right. I am merging to master and 1.6. Thanks @jerryshao for all the effort, and @zsxwing for getting this ready.

**Comment by jerryshao** [comment]
Thanks a lot @tdas and @zsxwing to make it merged, appreciated.

**Pull request closed without merging**